socs/agents/hwp_supervisor/agent.py

```diff
@@ -1104,45 +1106,56 @@ def check_acu_ok_for_spinup():
                 self.action.set_state(ControlState.Done(success=True))
 
             elif isinstance(state, ControlState.Brake):
+                pid_dir = int(query_pid_state()['direction'])
+                if state.fast:
+                    # set pcu fase brake state
+                    new_pcu_state = 'on_2' if (pid_dir == 1) else 'on_1'
+                    min_freq = 0.2
+                else:
+                    new_pcu_state = 'off'
+                    min_freq = 0.5
                 self.run_and_validate(
                     clients.pcu.send_command,
-                    kwargs={'command': 'off'}, timeout=None
+                    kwargs={'command': new_pcu_state}, timeout=None
                 )
-
-                # Flip PID direciton and tune stop
-                pid_dir = int(query_pid_state()['direction'])
-                new_d = '0' if (pid_dir == 1) else '1'
-                self.run_and_validate(clients.pid.set_direction,
-                                      kwargs=dict(direction=new_d))
-                self.run_and_validate(clients.pid.tune_stop)
-
+                if not state.fast:
+                    # Flip PID direciton and tune stop
+                    new_d = '0' if (pid_dir == 1) else '1'
+                    self.run_and_validate(clients.pid.set_direction,
+                                          kwargs=dict(direction=new_d))
+                    self.run_and_validate(clients.pid.tune_stop)
                 self.run_and_validate(clients.pmx.ign_ext)
                 self.run_and_validate(clients.pmx.set_v, kwargs={'volt': state.brake_voltage})
                 self.run_and_validate(clients.pmx.set_on)
 
                 time.sleep(10)
                 self.action.set_state(ControlState.WaitForBrake(
-                    min_freq=0.5,
+                    min_freq=min_freq,
                     prev_freq=hwp_state.enc_freq,
+                    fast=state.fast
                 ))
 
             elif isinstance(state, ControlState.WaitForBrake):
                 f0 = query_pid_state()['current_freq']
                 time.sleep(5)
                 f1 = query_pid_state()['current_freq']
-                if f0 < 0.5 or (f1 > f0):
+                if f0 < state.min_freq or (f1 > f0):
                     self.log.info("Turning off PMX and putting PCU in stop mode")
                     self.run_and_validate(clients.pmx.set_off)
                     self.run_and_validate(
                         clients.pcu.send_command,
                         kwargs={'command': 'stop'}, timeout=None
                     )
-                    self.action.set_state(ControlState.WaitForTargetFreq(
-                        target_freq=0,
-                        freq_tol=0.05,
-                        freq_tol_duration=30,
-                    ))
-                    return
+                    if state.fast:
+                        time.sleep(5)
+                        return
```

Comment:
This is not very safe. We should do WaitForTargetFreq.

But I think the motivation of this PR is to change the rotation direction quickly for special calibration, and you may not want to add WaitForTargetFreq. Therefore, it might be better to have a separate function that will change rotation direction, which will be faster than doing fast brake and spin up.

Reply:
Yes I agree. Just wondering what is the behavior if you PID to -2 Hz by changing the PID direction while sitting at 2 Hz? If we can get something like that to work I agree we should make it a specialized operation.

Reply:
> what is the behavior if you PID to -2 Hz by changing the PID direction while sitting at 2 Hz?

This command will also rapidly change the rotation direction. 
Changing PID direction flips the direction of acceleration and the HWP slows down. As the absolute PID frequency stays at 2 Hz, the PMX serves the highest voltage and current to recover 2 Hz, which results in a strong brake. Then the HWP speeds up to opposite direction and reaches -2 Hz.
But the current `hwp_supervisor` looks it doesn't expect this procedure so this will require some changes anyway.
Which looks better @ykyohei ?